### ISSUE TITLE
[Deployment Issue] Problem with real client IP within X-Forwarded-For header

### ISSUE THREAD

**Issue by alan-hadyk** [caprover/caprover#1077]
**What is the problem?**
I have a Caprover server running on DigitalOcean droplet. Inside Caprover, I run a Node.js server with Express. In order to send cookies back to client, I have to set `app.set("trust proxy", true)`. This option grabs left-most entry of `X-Forwarder-For` header ( https://expressjs.com/en/guide/behind-proxies.html ). However - `X-Forwarder-For` contains only 1 IP address, and it's not the client IP. Should I change something in my NGINX config to include the real IP? This is my current config (as you can see, I've been trying various tricks with `X-Forwarded-Proto` and  `Forwarded`) :

```
# Sets a $real_scheme variable whose value is the scheme passed by the load
# balancer in X-Forwarded-Proto (if any), defaulting to $scheme.
# Similar to how the HttpRealIp module treats X-Forwarded-For.
map $http_x_forwarded_proto $real_scheme {
  default $http_x_forwarded_proto;
  ''      $scheme;
}

map $remote_addr $proxy_forwarded_elem {
    # IPv4 addresses can be sent as-is
    ~^[0-9.]+$          "for=$remote_addr";

    # IPv6 addresses need to be bracketed and quoted
    ~^[0-9A-Fa-f:.]+$   "for=\"[$remote_addr]\"";

    # Unix domain socket names cannot be represented in RFC 7239 syntax
    default             "for=unknown";
}

map $http_forwarded $proxy_add_forwarded {
    # If the incoming Forwarded header is syntactically valid, append to it
    "~^(,[ \\t]*)*([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?(;([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?)*([ \\t]*,([ \\t]*([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?(;([!#$%&'*+.^_`|~0-9A-Za-z-]+=([!#$%&'*+.^_`|~0-9A-Za-z-]+|\"([\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|\\\\[\\t \\x21-\\x7E\\x80-\\xFF])*\"))?)*)?)*$" "$http_forwarded, $proxy_forwarded_elem";

    # Otherwise, replace it
    default "$proxy_forwarded_elem";
}

<%
if (s.forceSsl) {
%>
    server {

        listen       80;

        server_name  <%-s.publicDomain%>;

        # Used by Lets Encrypt
        location /.well-known/acme-challenge/ {
            root <%-s.staticWebRoot%>;
        }

        # Used by CapRover for health check
        location /.well-known/captain-identifier {
            root <%-s.staticWebRoot%>;
        }

        location / {
            return 302 https://$http_host$request_uri$is_args$query_string;
        }
    }
<%
}
%>


server {

    <%
    if (!s.forceSsl) {
    %>
        listen       80;
    <%
    }
    if (s.hasSsl) {
    %>
        listen              443 ssl http2;
        ssl_certificate     <%-s.crtPath%>;
        ssl_certificate_key <%-s.keyPath%>;
    <%
    }
    %>

        client_max_body_size 500m;

        server_name  <%-s.publicDomain%>;

        # 127.0.0.11 is DNS set up by Docker, see:
        # https://docs.docker.com/engine/userguide/networking/configure-dns/
        # https://github.com/moby/moby/issues/20026
        resolver 127.0.0.11 valid=10s;
        # IMPORTANT!! If you are here from an old thread to set a custom port, you do not need to modify this port manually here!!
        # Simply change the Container HTTP Port from the dashboard HTTP panel
        set $upstream http://<%-s.localDomain%>:<%-s.containerHttpPort%>;

        location / {

    <%
    if (s.httpBasicAuthPath) {
    %>
            auth_basic           "Restricted Access";
            auth_basic_user_file <%-s.httpBasicAuthPath%>; 
    <%
    }
    %>

            proxy_pass $upstream;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $real_scheme;
            proxy_set_header Forwarded $proxy_add_forwarded;

    <%
    if (s.websocketSupport) {
    %>
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
            proxy_http_version 1.1;
    <%
    }
    %>
        }

        # Used by Lets Encrypt
        location /.well-known/acme-challenge/ {
            root <%-s.staticWebRoot%>;
        }
        
        # Used by CapRover for health check
        location /.well-known/captain-identifier {
            root <%-s.staticWebRoot%>;
        }

        error_page 502 /captain_502_custom_error_page.html;
        location = /captain_502_custom_error_page.html {
                root <%-s.customErrorPagesDirectory%>;
                internal;
        }
}
```

**If applicable, content of `captain-definition` file:**

```
{
 "schemaVersion" :2 ,
 "dockerfilePath": "./Dockerfile"
}
```
`Dockerfile`:
```
FROM node:14.11.0

# Create app directory
WORKDIR /usr/src/app

# Copy package.json and yarn.lock
COPY package.json ./
COPY yarn.lock ./

# Install dependencies
RUN yarn
RUN yarn global add pm2
RUN yarn global add typescript

# Bundle app source
COPY . .

# Build app
RUN /usr/src/app/node_modules/typescript/bin/tsc -p /usr/src/app/tsconfig.json

# Prune devDependencies
ENV NODE_ENV=production
RUN yarn

ENV PORT 80
EXPOSE 80

CMD [ "yarn", "start" ]
```

**Steps to reproduce the problem:**
If possible, please try to recreate similar scenario with given `Dockerfile`, NGINX config and Express app (example generator -> https://github.com/seanpmaxwell/express-generator-typescript#readme ). 

**Answers to the following questions if applicable:**
- Your OS and version?
CapRover 1.9.0 on Ubuntu 18.04
- RAM?
1 GB memory (Basic / 1 GB / 1 vCPU droplet)

**Comment by githubsaturn** [comment]
Why did you change your nginx configs? The default nginx config passes through the correct client IP.

I've just tested this, and it prints the right IP address:
```
app.use(function(err, req, res, next) {
  console.log(req.headers["x-forwarded-for"]);
  next()
});
```

`x-forwarded-for` has only one IP, and that is the IP address of the client. I've tested using the default nginx config without any customization.

**Comment by alan-hadyk** [comment]
I changed the default nginx config, because IP was wrong with the default one. However - when I changed it back to the default one yesterday, it suddenly works! I have no idea why. I'm still trying to wrap my head around this. 😕

**Comment by delasource** [comment]
This is still up to date and a problem related with ipv6. ( #594 ? )
Try accessing a server via default ipv4 and the client ip is shown correctly (that is, in the X-Forwarded-For header, the X-Real-IP header, as well as in nginx logs!) But accessed via ipv6, the IP in the header is 172.18.0.1 (which is the ip of interface docker_gwbridge of the host machine).

I did have a custom global config because i need some SSL and log-path adjustments. But i can confirm that the default config does the exact same.

Running caprover 1.10.1 on Debian bullseye. The integrated nginx is version 1.23.1

I dont think that this is a caprover issue, maybe its Docker related.

---

**My usecase is this:**

I installed the Matomo one-click-app for analytics (formerly piwik). But with the above problem, every ipv6-visitor is recorded as 172.18.0.1.
You see where the problems begin?
